File: android/src/main/kotlin/com/nghinv/device_info/NDeviceInfoPlugin.kt
package com.nghinv.device_info

import android.content.Context
import android.os.Build
import android.provider.Settings
import androidx.annotation.NonNull

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.util.Locale

class NDeviceInfoPlugin: FlutterPlugin, MethodCallHandler {
  private lateinit var channel : MethodChannel
  private lateinit var context: Context

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "n_device_info")
    context = flutterPluginBinding.applicationContext
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    if (call.method == "getDeviceInfo") {
      val appVersion = context.packageManager.getPackageInfo(context.packageName, 0).versionName
      val info = mapOf(
        "id" to getImei(),
        "name" to getDeviceName(),
        "isSimulator" to isEmulator,
        "version" to Build.VERSION.RELEASE,
        "buildVersion" to appVersion,
        "model" to Build.MODEL,
        "manufacturer" to Build.MANUFACTURER,
        "os" to "Android",
        "osVersion" to Build.VERSION.RELEASE,
        "userAgent" to "Android ${Build.VERSION.RELEASE}"
      )

      result.success(info)
    } else {
      result.notImplemented()
    }
  }

  private fun getDeviceName(): String {
    val manufacturer = Build.MANUFACTURER
    val model = Build.MODEL
    return if (model.startsWith(manufacturer)) {
      model.replaceFirstChar { if (it.isLowerCase()) it.titlecase(Locale.ROOT) else it.toString() }
    } else {
      manufacturer.replaceFirstChar { if (it.isLowerCase()) it.titlecase(Locale.ROOT) else it.toString() } + " " + model
    }
  }

  private fun getImei(): String {
    return try {
      Settings.Secure.getString(context.contentResolver, Settings.Secure.ANDROID_ID)
    } catch (e: Exception) {
      Build.ID
    }
  }

  private val isEmulator: Boolean by lazy {
    return@lazy (Build.BRAND.startsWith("generic") && Build.DEVICE.startsWith("generic")
            || Build.FINGERPRINT.startsWith("generic")
            || Build.FINGERPRINT.startsWith("unknown")
            || Build.HARDWARE.contains("goldfish")
            || Build.HARDWARE.contains("ranchu")
            || Build.MODEL.contains("google_sdk")
            || Build.MODEL.contains("Emulator")
            || Build.MODEL.contains("Android SDK built for x86")
            || Build.MANUFACTURER.contains("Genymotion")
            || Build.PRODUCT.contains("sdk_google")
            || Build.PRODUCT.contains("google_sdk")
            || Build.PRODUCT.contains("sdk")
            || Build.PRODUCT.contains("sdk_x86")
            || Build.PRODUCT.contains("vbox86p")
            || Build.PRODUCT.contains("emulator")
            || Build.PRODUCT.contains("simulator"))
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
}

File: ios/Classes/NDeviceInfoPlugin.swift
import Flutter
import UIKit

public class NDeviceInfoPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "n_device_info", binaryMessenger: registrar.messenger())
    let instance = NDeviceInfoPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    if (call.method == "getDeviceInfo") {
      let device = UIDevice.current
      let info: [String: Any] = [
        "id": device.identifierForVendor?.uuidString,
        "name": device.name,
        "isSimulator": device.name.contains("Simulator"),
        "version": device.systemVersion,
        "buildVersion": Bundle.main.infoDictionary?["CFBundleVersion"] as? String,
        "model": device.model,
        "manufacturer": "Apple",
        "os": "iOS",
        "osVersion": device.systemVersion,
        "userAgent": "iOS"
      ]

      result(info)
    } else {
      result(FlutterMethodNotImplemented)
    }
  }
}
